torch/csrc/jit/tensorexpr/cuda_half_support.h

```diff
@@ -85,12 +85,32 @@ class CudaHalfScalarRewriter : public IRMutator {
   const Expr* mutate(const Var* v) override {
     auto it = var_map.find(v);
     if (it != var_map.end()) {
-      return it->second;
+      return new Cast(kHalf, it->second);
     }
 
     return v;
   }
 
+  const Expr* mutate(const Cast* v) override {
+    const Expr* child = v->src_value()->accept_mutator(this);
+
+    const Cast* cast_child = dynamic_cast<const Cast*>(child);
+    if (cast_child) {
+      if ((v->dtype().scalar_type() == ScalarType::Float ||
```

Comment:
Ah, not all casts are safe to eliminate - e.g `Float(Int(X))` is not safe to replace with `X`. Rather than work out what the correct behavior in all cases I decided to keep this limited to float and half since we're casting all halfs in this backend anyway.

Reply:
Hm, i may have misread this I read this as int(float(int_val)) - i think any cast from a type1 -> type2  -> type1 can always be removed right ?

Reply:
I don't think so, `Float(Int(X))` isn't safe when X is Float, neither is `Int(Float(Int(X)))` when X is a floating type.